src/groups.py

```diff
@@ -0,0 +1,380 @@
+        """
+        endpoint = ENDPOINT_MAPPING[member_type]
+        url = f"{RANGER_URL}/service/xusers/{endpoint}"
+        response = requests.get(
+            url, headers=HEADERS, auth=self._auth, timeout=10
+        )
+        j = json.loads(response.text)
+        return j
+
+    @handle_service_error
+    @retry(max_retries=3, delay=2, backoff=2)
+    def _delete_members(self, member_type, value):
+        """Send a DELETE request to the Ranger API for a member.
+
+        Args:
+            member_type: The type of Ranger member (group, user or membership).
+            value: The identifying value of the member to delete.
+        """
+        if member_type == "group" and value in SYSTEM_GROUPS:
+            return
+
+        if member_type == "membership" and value[0] in SYSTEM_GROUPS:
+            return
+
+        ids = self.charm._state.id_mapping[member_type]
+        value_id = ids[str(value)]
+
+        endpoint = ENDPOINT_MAPPING[member_type]
+        url = f"{RANGER_URL}/service/xusers/{endpoint}/{value_id}"
+        response = requests.delete(
+            url, headers=HEADERS, auth=self._auth, timeout=10
+        )
+
+        if response.status_code == 204:
+            logger.info(f"Deleted {member_type}: {value_id}")
+        else:
+            logger.info(f"Unable to delete {member_type}: {value_id}")
+
+    @handle_service_error
+    @retry(max_retries=3, delay=2, backoff=2)
+    def _create_members(self, member_type, fields, value):
+        """Send a POST request to create a member in the Ranger API.
+
+        Args:
+            member_type: The type of Ranger member (group, user or membership).
+            fields: Fields required for creating the payload.
+            value: The identifying value for the member. Name or Tuple.
+        """
+        payload = create_payload(member_type, fields)
+        endpoint = ENDPOINT_MAPPING[member_type]
+        url = f"{RANGER_URL}/service/xusers/{endpoint}"
+
+        response = requests.post(
+            url,
+            headers=HEADERS,
+            json=payload,
+            auth=self._auth,
+            timeout=10,
+        )
+        if response.status_code == 200:
+            logger.info(f"Created {member_type}: {value}")
+        else:
+            logger.info(f"Unable to create {member_type}: {value}")
+
+    @handle_service_error
+    def _transform_apply_values(self, data, member_type):
+        """Get list of users, groups or memberships to apply from configuration file.
+
+        Args:
+            data: User, group or membership data.
+            member_type: The type of Ranger member (group, user or membership).
+
+        Returns:
+            List of users, groups or memberships to apply.
+        """
+        if member_type in ["user", "group"]:
+            values = [member["name"] for member in data]
+            return values
+
+        _ = self._get_existing_values("user")
+        _ = self._get_existing_values("group")
```

Comment:
do you really need these 2 lines here? the users and groups id mapping should be loaded at this point no?

Reply:
Yes, the id mapping would have been loaded for all the existing users, not the users that were created new. Let me know if you can think of a nicer way to achieve this.

Reply:
but why though, isn't the sync called first for users, then groups, then memberships? I guess you come here only after calling sync 2 times no?

Reply:
At the beginning of the "user" sync, we do an API call to get all existing users and add them to _state mapping the user name to id. We then create new users (for which we do not have the id's), then we do groups, this just gets the existing groups and does this id mapping again, it then adds the new groups (for which we do not have stored the ids). Then it comes to memberships, and now if we need to associate a new user with any group we will have a problem as we do not have the new user id. Annoyingly it is not provided when creating the user, or we could just add it to the _state dictionary.

Reply:
never mind, got it, i think you should update that when you create the new members

Reply:
Updated this can you take another look?

Reply:
Any specific reason you are loading them into `_`? You can just call them.

Reply:
You are right, this has been removed in favour of another method of getting the user id's